File: Step 3 - Arrays/Lec 1 - Easy/Program_09_optimal.cpp
#include <bits/stdc++.h>
using namespace std;

// Find the union of two sorted arrays.
// 2-pointer approach

int main() {
    int n1, n2;
    cin >> n1;
    vector<int> a(n1);
    for (int i = 0; i < n1; i++) {
        cin >> a[i];
    }
    cin >> n2;
    vector<int> b(n2);
    for (int i = 0; i < n2; i++) {
        cin >> b[i];
    }

    vector<int> unionArr;
    int i = 0, j = 0;
    while (i < n1 && j < n2) {
        if (a[i] <= b[j]) {
            if (unionArr.size() == 0 || unionArr.back() != a[i]) {
                unionArr.push_back(a[i]);
            }
            i++;
        } else {
            if (unionArr.size() == 0 || unionArr.back() != b[j]) {
                unionArr.push_back(b[j]);
            }
            j++;
        }
    }

    while (i < n1) {
        if (unionArr.size() == 0 || unionArr.back() != a[i]) {
            unionArr.push_back(a[i]);
        }
        i++;
    }
    while (j < n2) {
        if (unionArr.size() == 0 || unionArr.back() != b[j]) {
            unionArr.push_back(b[j]);
        }
        j++;
    }

    for (auto element : unionArr) {
        cout << element << " ";
    }
}

// Time-complexity:  O(n1+n2)
// Space-complexity: O(n1+n2) for returning the unionArr

File: Step03_Arrays/Lec1_Easy/Program_09_optimal.java
package Step03_Arrays.Lec1_Easy;

import java.util.*;

// Find the union of two sorted arrays.
// 2-pointer approach

public class Program_09_optimal {
    public static ArrayList<Integer> getUnion(int[] arr1, int[] arr2) {
        int n1 = arr1.length;
        int n2 = arr2.length;

        ArrayList<Integer> union = new ArrayList<>();
        int i = 0, j = 0;

        while (i < n1 && j < n2) {
            if (arr1[i] <= arr2[j]) {
                if (union.size() == 0 || union.get(union.size() - 1) != arr1[i]) {
                    union.add(arr1[i]);
                }
                i++;
            } else {
                if (union.size() == 0 || union.get(union.size() - 1) != arr2[j]) {
                    union.add(arr2[j]);
                }
                j++;
            }
        }

        while (i < n1) {
            if (union.size() == 0 || union.get(union.size() - 1) != arr1[i]) {
                union.add(arr1[i]);
            }
            i++;
        }
        while (j < n2) {
            if (union.size() == 0 || union.get(union.size() - 1) != arr2[j]) {
                union.add(arr2[j]);
            }
            j++;
        }

        return union;
    }

    public static void main(String[] args) {
        int[] arr1 = { 1, 3, 3, 7, 9, 9, 9 };
        int[] arr2 = { 2, 4, 4, 4, 6, 8, 8, 8, 10 };

        for (int element : getUnion(arr1, arr2)) {
            System.out.print(element + " ");
        }
    }
}

// Time-complexity : O(n1+n2)
// Space-complexity : O(n1+n2) for returning the union
